File: Week4_Arrays/squares-of-sorted-array-eff.java
class Solution {
    public int[] sortedSquares(int[] nums) {
        int[] ans = new int[nums.length];   
        /* Assign Pointers to add new value to array */
        int ptr = nums.length - 1;
        /* Pointer for most negative and positive value*/
        int l = 0;
        int r = nums.length - 1;
        
        /* While new Array not completely populated */
        while (ptr >= 0) {
            /* Compare Absolute value */
            if (Math.abs(nums[l]) > Math.abs(nums[r])) {
                ans[ptr] = nums[l] * nums[l];
                l++;
            } else {
                ans[ptr] = nums[r] * nums[r];
                r--;
            }
            /* Move ptr to left to assign new value */
            ptr--;
        }
        return ans;
    }
}

File: Session2_Arrays/squares-of-sorted-array-eff.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        # Assign new array
        res = [0] * len(nums)
        # Assign left and right forward for
        # Most negative and Most positive values
        l = 0
        r = len(nums) - 1
        # Assign pointer to last index since the square
        # of a most negative/ positive value would be
        # Highest in the new array
        ptr = len(nums) - 1
        # While new array is not completely populated
        while ptr >= 0:
            # Compare absolute magnitude
            if abs(nums[l]) > abs(nums[r]):
                res[ptr] = nums[l] ** 2
                l += 1
            else:
                res[ptr] = nums[r] ** 2
                r -= 1
            # Shift pointer for next element to be added
            ptr -= 1
            
        return res
        